tests/test_pytorch/test_fsdp.py

```diff
@@ -0,0 +1,791 @@
+# Copyright (c) 2023 Habana Labs, Ltd. an Intel Company
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+from functools import partial
+from pathlib import Path
+from typing import Optional, Tuple
+from unittest import mock
+from unittest.mock import ANY, MagicMock, Mock
+
+import pytest
+import torch
+import torch.nn as nn
+from lightning_utilities import module_available
+from torch.distributed.fsdp.fully_sharded_data_parallel import CPUOffload, FullyShardedDataParallel, MixedPrecision
+from torch.distributed.fsdp.wrap import size_based_auto_wrap_policy, wrap
+from torchmetrics import Accuracy
+
+if module_available("lightning"):
+    from lightning.fabric.utilities.imports import (
+        _TORCH_GREATER_EQUAL_2_0,
+        _TORCH_GREATER_EQUAL_2_1,
+    )
+    from lightning.pytorch import Trainer, seed_everything
+    from lightning.pytorch.demos import LightningTransformer
+    from lightning.pytorch.demos.boring_classes import BoringDataModule, BoringModel
+elif module_available("pytorch_lightning"):
+    from lightning_fabric.utilities.imports import (
+        _TORCH_GREATER_EQUAL_2_0,
+        _TORCH_GREATER_EQUAL_2_1,
+    )
+    from pytorch_lightning import Trainer
+    from pytorch_lightning.demos import LightningTransformer
+    from pytorch_lightning.demos.boring_classes import BoringDataModule, BoringModel
+
+from lightning_habana.pytorch.accelerator import HPUAccelerator
+from lightning_habana.pytorch.plugins.fsdp_precision import HPUFSDPPrecision, HPUPrecisionPlugin
+from lightning_habana.pytorch.strategies import HPUDDPStrategy, HPUFSDPStrategy
+
+
+class TestFSDPModel(BoringModel):
+    def __init__(self):
+        super().__init__()
+        self.layer: Optional[nn.Module] = None
+
+    def _init_model(self) -> None:
+        self.layer = torch.nn.Sequential(torch.nn.Linear(32, 32), torch.nn.ReLU(), torch.nn.Linear(32, 2))
+
+    def configure_model(self) -> None:
+        if self.layer is None:
+            self._init_model()
+        # the model is already wrapped with FSDP: no need to wrap again!
+        if isinstance(self.layer, FullyShardedDataParallel):
+            return
+
+        self.layer = wrap(self.layer)
+
+    def configure_optimizers(self):
+        # There is some issue with SGD optimizer state in FSDP
+        return torch.optim.AdamW(self.layer.parameters(), lr=0.1)
+
+    def on_train_batch_start(self, batch, batch_idx):
+        assert batch.dtype == torch.float32
```

Comment:
self._Assert ?

Reply:
I think the check doesn't apply at this place. The test is taken as is from lightning upstream. https://github.com/Lightning-AI/pytorch-lightning/blob/c235f20e7131af2c7be4cc9080d3c946d93d58ea/tests/tests_pytorch/strategies/test_fsdp.py#L66